bindings/profiler.cc

```diff
@@ -101,61 +105,226 @@ CpuProfiler* cpuProfiler = CpuProfiler::New(v8::Isolate::GetCurrent());
-  js_node->Set(Nan::New<String>("columnNumber").ToLocalChecked(),
-    Nan::New<Integer>(node->GetColumnNumber()));
-  js_node->Set(Nan::New<String>("hitCount").ToLocalChecked(),
-    Nan::New<Integer>(node->GetHitCount()));
   int32_t count = node->GetChildrenCount();
   Local<Array> children = Nan::New<Array>(count);
   for (int32_t i = 0; i < count; i++) {
     children->Set(i, TranslateTimeProfileNode(node->GetChild(i)));
   }
-  js_node->Set(Nan::New<String>("children").ToLocalChecked(),
-    children);
-  return js_node;
+
+  return CreateTimeNode(
+    node->GetFunctionName(),
+    node->GetScriptResourceName(),
+    Nan::New<Integer>(node->GetScriptId()),
+    Nan::New<Integer>(node->GetLineNumber()),
+    Nan::New<Integer>(node->GetColumnNumber()),
+    Nan::New<Integer>(node->GetHitCount()),
+    children
+  );
 }
 
-Local<Value> TranslateTimeProfile(const CpuProfile* profile) {
+Local<Value> TranslateTimeProfile(const CpuProfile* profile, bool hasDetailedLines) {
   Local<Object> js_profile = Nan::New<Object>();
   js_profile->Set(Nan::New<String>("title").ToLocalChecked(),
     profile->GetTitle());
-  js_profile->Set(Nan::New<String>("topDownRoot").ToLocalChecked(),
+
+#if NODE_MODULE_VERSION > NODE_11_0_MODULE_VERSION
+  if (hasDetailedLines) {
+    js_profile->Set(
+      Nan::New<String>("topDownRoot").ToLocalChecked(),
+      TranslateLineNumbersTimeProfileRoot(profile->GetTopDownRoot()));
+  } else {
+    js_profile->Set(
+      Nan::New<String>("topDownRoot").ToLocalChecked(),
+      TranslateTimeProfileNode(profile->GetTopDownRoot()));
+  }
+#else
+  js_profile->Set(
+    Nan::New<String>("topDownRoot").ToLocalChecked(),
     TranslateTimeProfileNode(profile->GetTopDownRoot()));
+#endif
   js_profile->Set(Nan::New<String>("startTime").ToLocalChecked(),
     Nan::New<Number>(profile->GetStartTime()));
   js_profile->Set(Nan::New<String>("endTime").ToLocalChecked(),
     Nan::New<Number>(profile->GetEndTime()));
   return js_profile;
 }
 
+// Signature:
+// startProfiling(runName: string, includeLineInfo: boolean)
 NAN_METHOD(StartProfiling) {
-  Local<String> name = info[0].As<String>();
+  if (info.Length() != 2) {
+    return Nan::ThrowTypeError("StartProfling must have two arguments.");
+  }
+  if (!info[0]->IsString()) {
+    return Nan::ThrowTypeError("First argument must be a string.");
+  }
+  if (!info[1]->IsBoolean()) {
+    return Nan::ThrowTypeError("Second argument must be a boolean.");
+  }
+
+  Local<String> name =
+      Nan::MaybeLocal<String>(info[0].As<String>()).ToLocalChecked();
 
-  // Sample counts and timestamps are not used, so we do not need to record
-  // samples.
-  cpuProfiler->StartProfiling(name, false);
+// Sample counts and timestamps are not used, so we do not need to record
+// samples.
+bool recordSamples = false;
+
+// Line-level accurate line information is not available in Node 11 or earlier.
+#if NODE_MODULE_VERSION > NODE_11_0_MODULE_VERSION
+  bool includeLineInfo =
```

Comment:
terminology: what is the difference between "has" used above and "include" used here? Is this intentionally different?

Reply:
Switched to using "includeLineInfo" (for starting the profile, when we're requesting line information), and "includedLineInfor" (where it indicates that line info is present).

https://github.com/google/pprof-nodejs/pull/73/files